### ISSUE TITLE
Dryrun endpoint does not recognize app create transactions

### ISSUE THREAD

**Issue by jdtzmn** [algorand/go-algorand#2307]
### Subject of the issue

The dryrun endpoint does not handle app create transactions correctly. Submitting a dryrun request with a single app create transaction "fails" with `uploaded state did not include app id 0 referenced in txn[0]`.

### Your environment

A local network inside of an Algorand sandbox.

### Steps to reproduce

1. Bring up an algod node
2. Send an app create transaction to the dryrun endpoint 

The following code serves as a template to send a simple app create transaction to the algod v2 dryrun endpoint.

```js
import algosdk from 'algosdk';
const client = new algosdk.Algodv2(...);
const account = algosdk.generateAccount();

const suggestedParams = await client.getTransactionParams().do();
const trivialProgram = Uint8Array.from([2, 20, 1, 1, 22]); // int 1

const appCreateTxn = algosdk.makeApplicationCreateTxnFromObject({
    from: account.addr,
    onComplete: algosdk.OnApplicationComplete.OptInOC,
    approvalProgram: trivialProgram,
    clearProgram: trivialProgram,
    suggestedParams,
    numGlobalByteSlices: 0,
    numGlobalInts: 0,
    numLocalByteSlices: 0,
    numLocalInts: 0,
});

// Dryrun logic
const options = {
    txns: [{ txn: appCreateTxn }]
};
const dryrunRequest = new algosdk.modelsv2.DryrunRequest(options);
const dryrunResponse = await client.dryrun(dryrunRequest).do();
console.log(dryrunResponse.txns);
```

### Expected behaviour

Dryrun endpoint should recognize an app create transaction and not require the uploaded state to include an app with an app id of 0.

### Actual behaviour

The dryrun request fails with `uploaded state did not include app id 0 referenced in txn[0]`. I would assume it is trying to get the approval program from state and doesn't recognize an app id of 0 as an app create transaction.

**Comment by jdtzmn** [comment]
After some code searching, app creation does seem to be supported. Is this behavior documented anywhere?

https://github.com/algorand/go-algorand/blob/06faea5ac76767b4209e8f09ee3b5ac7231d2fb1/daemon/algod/api/server/v2/dryrun.go#L431-L437

**Comment by jdtzmn** [comment]
I took some notes regarding the dryrun interface and the dryrun request structure while I was troubleshooting that I thought I would include here as a reference. Hopefully these will help otherws with similar issues in the future.

- Dryrun sources must be the actual source string and not the compiled binary (`Uint8Array` in JavaScript).
- Dryrun app approval program through application params works with raw byte string `\0x00\0x01` of a compiled program and are _not_ base64 encoded. In JavaScript, this means using the `Buffer.toString()` method with no encoding.
- Dryrun requests for an app create call seem to look for the approval program in the following order: sources, then app params, but never the transactions themselves (including the `approvalProgram` in an app create transaction never seems to work). This also means that included sources "overrule" app params if both are included.
- Foreign assets *do* work with dryrun requests.

**Comment by jdtzmn** [comment]
I'm closing this as many of the issues included here were solved by #2448.

**Comment by fionnachan** [comment]
Adding onto the observations above, the error `uploaded state did not include app id 0 referenced in txn[0]` will go away after passing `apps: [app]` into `DryrunRequest`, but `app` of type `Application` must have a **non zero** id, or dryrun will throw the following error: `['ApprovalProgram', 'REJECT', 'cannot make logic ledger for app index 0']`.
